### PR TITLE
RM-60271 configurable draft sequence

### PR DESCRIPTION
**Pull request by mjidox**

### CODE REVIEW

File: changelogs/unreleased/60271.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+title: "Sequence: configurable draft sequence"
```
Comment (dtr-axelor) [inline changelogs/unreleased/60271.yml:2]:
```suggestion
title: "Enable configuration of draft sequence prefix"
```

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 2]
```diff
@@ -58,7 +58,7 @@
 @Singleton
 public class SequenceService {
 
-  protected static final String DRAFT_PREFIX = "#";
+  protected static String DRAFT_PREFIX = "#";
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:61]: why?

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 3]
```diff
@@ -314,6 +314,17 @@ public String getDefaultTitle(Sequence sequence) {
     return item.getTitle();
   }
 
+  /**
+   * Get draft sequence number prefix.
+   *
+   * @return
+   */
+  public String getDraftPrefix() {
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:322]:
```suggestion
  protected String getDraftPrefix() {
```

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 4]
```diff
@@ -314,6 +314,17 @@ public String getDefaultTitle(Sequence sequence) {
...
+    if (appBaseService.getAppBase().getDraftSuffix() != null)
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:323]: Use optional pattern, for example:
 `return Optional.ofNullable(appBaseService.getAppBase().getDraftSuffix()).orElse(DRAFT_PREFIX);`

**Review by dtr-axelor — CHANGES_REQUESTED** [review]

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 5]
```diff
@@ -328,14 +338,15 @@ public String getDraftSequenceNumber(Model model) throws AxelorException {
...
+    String draftPrefix = getDraftPrefix();
+    return String.format("%s%d", draftPrefix, model.getId());
   }
 
   /**
    * Get draft sequence number with leading zeros.
    *
    * @param model
-   * @param padding
+   * @param zeroPadding
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:349]:
```suggestion
   * @param padding
```
unrelated change

File: axelor-base/src/main/resources/domains/AppBase.xml [hunk 6]
```diff
@@ -20,6 +20,9 @@
     <!-- Interfaces QAS et SVI -->
     <string name="qasWsdlUrl" title="URL to QAS WSDL"/>
 
+
+    <string name="draftPrefix" title="Draft prefix"/>
```
Comment (dtr-axelor) [inline axelor-base/src/main/resources/domains/AppBase.xml:24]:
```suggestion
    <string name="draftPrefix" title="Draft prefix" default="#"/>
```
Just a small suggestion because I think it would be a bit more user friendly this way. Otherwise, the implementation seems fine, good job! 👍🏽

**Review by dtr-axelor — CHANGES_REQUESTED** [review]

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 7]
```diff
@@ -403,4 +415,12 @@ public List<SequenceVersion> updateSequenceVersions(
 
     return sequenceVersionList;
   }
+
+  public void validateSequence(Sequence sequence) throws AxelorException {
+    if (sequence.getPrefixe().startsWith(appBaseService.getAppBase().getDraftPrefix()))
+      throw new AxelorException(
+          appBaseService.getAppBase(),
+          TraceBackRepository.CATEGORY_CONFIGURATION_ERROR,
+          I18n.get("Sequence prefix cannot contain draft prefix"));
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:424]: - [ ] Error messages should be centralized in BaseExceptionMessage.
- [ ] Add French translation
- [ ] Also change the implementation as follows:

```suggestion
    String draftPrefix = getDraftPrefix();
    if (sequence.getPrefixe().startsWith(draftPrefix))
      throw new AxelorException(
          sequence,
          TraceBackRepository.CATEGORY_CONFIGURATION_ERROR,
          I18n.get("Sequence prefix cannot start with the draft prefix '%s'."),
          draftPrefix);
```

File: axelor-base/src/main/resources/views/Sequence.xml [hunk 8]
```diff
@@ -122,11 +123,17 @@
     <error message="Prefix or suffix must not contain characters forbidden in filenames."
       if="p = ~/^[^\/\\?*:\|&quot;&lt;>]*$/; prefixe &amp;&amp; !(prefixe ==~ p) || suffixe &amp;&amp; !(suffixe ==~ p)"/>
     <error message="Prefix cannot start with #." if="prefixe ==~ /^#.*$/"/>
+    <error message="Prefix cannot start with the prefix draft."
+      if="prefixe ==__config__.app.getApp('base').getDraftPrefix()"/>
+
```
Comment (dtr-axelor) [inline axelor-base/src/main/resources/views/Sequence.xml:128]: why this change if you already do it on java?

File: axelor-base/src/main/resources/views/Sequence.xml [hunk 9]
```diff
@@ -122,11 +123,17 @@
...
   </action-validate>
 
   <action-method name="action-method-sequence-document-concerned-onchange">
     <call class="com.axelor.apps.base.web.SequenceController" method="getDefaultTitle"/>
   </action-method>
+  <action-method name="action-method-sequence-validate-before-save">
```
Comment (dtr-axelor) [inline axelor-base/src/main/resources/views/Sequence.xml:134]:
```suggestion

  <action-method name="action-method-sequence-validate-before-save">
```
for readability

File: axelor-base/src/main/resources/views/AppBase.xml [hunk 10]
```diff
@@ -36,8 +36,11 @@
         <field name="unitHours" form-view="unit-form" grid-view="unit-grid"/>
         <field name="unitMinutes" form-view="unit-form" grid-view="unit-grid"/>
         <field name="teamManagement" widget="boolean-switch"/>
-        <field name="manageSubsidiaryPartner" widget="boolean-switch"/>
-        <field name="checkExistingSequenceOnGeneration" widget="boolean-switch"/>
+        <field name="manageSubsidiaryPartner" widget="boolean-switch" colSpan="4"/>
+        <field name="checkExistingSequenceOnGeneration" widget="boolean-switch" colSpan="4"/>
+        <field name="draftPrefix" colSpan="4"/>
+
+
```
Comment (dtr-axelor) [inline axelor-base/src/main/resources/views/AppBase.xml:43]:
```suggestion
        <field name="manageSubsidiaryPartner" widget="boolean-switch" />
        <field name="checkExistingSequenceOnGeneration" widget="boolean-switch" colSpan="3"/>
        <field name="draftPrefix" colSpan="3"/>
```

We should avoid changing too much the UI

**Review by dtr-axelor — CHANGES_REQUESTED** [review]

File: axelor-base/src/main/java/com/axelor/apps/base/web/SequenceController.java [hunk 11]
```diff
@@ -42,6 +43,17 @@ public void getDefaultTitle(ActionRequest request, ActionResponse response) {
...
 
+  public void validateSequence(ActionRequest request, ActionResponse response) {
+    Sequence sequence = request.getContext().asType(Sequence.class);
+    if (!Strings.isNullOrEmpty(sequence.getCodeSelect())) {
+      try {
+        Beans.get(SequenceService.class).validateSequence(sequence);
+      } catch (AxelorException e) {
+        TraceBackService.trace(response, e, ResponseMessageType.ERROR);
+      }
+    }
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/web/SequenceController.java:54]:
```suggestion
      try {
        Beans.get(SequenceService.class).validateSequence(sequence);
      } catch (AxelorException e) {
        TraceBackService.trace(response, e, ResponseMessageType.ERROR);
      }
```

File: axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java [hunk 12]
```diff
@@ -403,4 +415,16 @@ public List<SequenceVersion> updateSequenceVersions(
 
     return sequenceVersionList;
   }
+
+  @Transactional(rollbackOn = Exception.class)
```
Comment (dtr-axelor) [inline axelor-base/src/main/java/com/axelor/apps/base/service/administration/SequenceService.java:419]:
```suggestion
```

**Review by dtr-axelor — CHANGES_REQUESTED** [review]

**Comment by ale-axelor** [comment]
see #11071
